File: auth_samples_flutter/lib/jwt.dart
import 'dart:convert';

Map<String, dynamic>? decodeJwt(String? token) {
  if (token == null || token.isEmpty) return null;
  final parts = token.split('.');
  if (parts.length < 2) return null;
  try {
    final payload = utf8.decode(base64Url.decode(base64Url.normalize(parts[1])));
    return json.decode(payload) as Map<String, dynamic>;
  } catch (_) {
    return null;
  }
}

String formatUnixTime(int? seconds) {
  if (seconds == null || seconds <= 0) return 'unknown time';
  final date = DateTime.fromMillisecondsSinceEpoch(seconds * 1000);
  return date.toLocal().toString();
}

File: auth_samples_angular/src/app/utils/jwt.ts
export function decodeJwt(token: string): Record<string, unknown> | null {
  if (!token) return null;
  const parts = token.split(".");
  if (parts.length < 2) return null;

  try {
    const payload = decodeBase64Url(parts[1]);
    return JSON.parse(payload) as Record<string, unknown>;
  } catch {
    return null;
  }
}

export function formatUnixTime(unixSeconds: number | undefined) {
  if (!unixSeconds) return "unknown time";
  const date = new Date(unixSeconds * 1000);
  return date.toLocaleString();
}

function decodeBase64Url(value: string) {
  const padded = value.replace(/-/g, "+").replace(/_/g, "/");
  const padLength = padded.length % 4;
  const base64 = padLength ? padded.padEnd(padded.length + (4 - padLength), "=") : padded;
  return atob(base64);
}
